ssm/lds.py

```diff
@@ -744,26 +904,104 @@ def _fit_laplace_em(self, variational_posterior, datas,
+                    # expected state or covariance calculation.
+                    log_Z_dyn, log_Z_ini, log_Z_obs = 0, 0, 0
+                    log_Z, smoothed_mus, smoothed_sigmas, ExxnT = \
+                        kalman_info_smoother(
+                            prms["J_ini"], prms["h_ini"], log_Z_ini,
+                            prms["J_dyn_11"], prms["J_dyn_21"],
+                            prms["J_dyn_22"], prms["h_dyn_1"], prms["h_dyn_2"], log_Z_dyn,
+                            prms["J_obs"], prms["h_obs"], log_Z_obs
+                        )
+                    Ex.append(smoothed_mus)
+                    mumuT = np.swapaxes(smoothed_mus[:, None], 2,1) @ smoothed_mus[:, None]
+                    ExxT.append(smoothed_sigmas + mumuT)
+                    ExyT.append(ExxnT)
+
+                self._fit_laplace_em_params_update_exact_mstep(
+                    discrete_expectations,
+                    # In the exact, the sample of the continuous sates
+                    # won't be needed, but we pass it for consistency.
+                    continuous_samples,
+                    datas,
+                    inputs,
+                    masks,
+                    tags,
+                    emission_optimizer,
+                    emission_optimizer_maxiter,
+                    alpha,
+                    continuous_expectations=(Ex, ExxT, ExyT),
+                    )
+
             # Default is partial M-step given a sample from q(x)
-            if learning and parameters_update=="mstep":
-                self._fit_laplace_em_params_update(
-                    discrete_expectations, continuous_expectations, datas, inputs, masks, tags,
+            elif learning and parameters_update=="stoch_mstep":
+                self._fit_laplace_em_params_update_stoch_mstep(
+                    discrete_expectations, continuous_samples, datas, inputs, masks, tags,
                     emission_optimizer, emission_optimizer_maxiter, alpha)
+
             # Alternative is SGD on all parameters with samples from q(x)
             elif learning and parameters_update=="sgd":
                 self._fit_laplace_em_params_update_sgd(
                     variational_posterior, datas, inputs, masks, tags,
                     emission_optimizer="adam",
                     emission_optimizer_maxiter=emission_optimizer_maxiter)
 
-            # 4. Compute ELBO
-            elbo = self._laplace_em_elbo(variational_posterior, datas, inputs, masks, tags)
+            elif learning:
+                raise ValueError("Invalid argument for parameters_update: {}. " \
+                    "Must be one of: stoch_mstep, exact_mstep, sgd.".format(parameters_update))
+
+            # 4. Compute ELBO. For now, we are wasting a lot of computation
+            # by re-running the Kalman smoother. This is just a first pass to 
+            # get functionality. TODO: Fix!!
+            Ex, ExxT, ExyT, log_Zs = [], [], [], []
```

Comment:
This whole block should be replaced with
```elbo = self._laplace_em_elbo(variational_posterior, datas, inputs, masks, tags)```
We don't need to compute any expectations in this top level function.  In fact, we don't need to compute these expectations in `self._laplace_em_elbo` either, since the `variational_posterior` has access to them.  See my comment above.